### DISCUSSION TITLE
Why push data to array on async/await?

### DISCUSSION THREAD

**Discussion by jiho5993** [nodejs/help#4973] [Q&A] [answered]
in func.js
```javascript
const p = (res) => {
    return new Promise((resolve, reject) => {
        res.push('2');
        resolve(res);
    });
};
```
in main.js
```javascript
const run = async () => {
    const res = [];
    res.push('1');
    
    console.log('before...');
    for (const item of res) console.log(item);
    
    try {
        await p(res); // call p func in func.js
        
        console.log('after...');
        for (const item of res) console.log(item); // ???
    } catch (e) {
        console.log('ERROR');
    }
};

run();
/** result
 * before...
 * 1
 * after...
 * 1
 * 2
 */
```

```func.js``` declare promise function and ```main.js``` run promise function using async/await.

Why push 2 to res array if i don't assign returned new array?
Does the resolve function have the same property as the pointer of the C language?

**Comment by palamarchukser** [comment, marked as answer]
it happens because you are passing the link to an array. to avoid this behaviour
`await p([...res]); // call p func in func.js`

```
before...
1
after...
1
```

**Reply by jiho5993** [reply]
thank you!
So, Is the execution result of the code I uploaded like that?
or I want to know keyword related to this.

**Reply by palamarchukser** [reply]
didn't really understand the question
just change, from `await p(res);` to `await p([...res]);`

**Reply by himself65** [reply]
Because an array in js is a reference to an object(like a pointer in C), in your code, you passed into a reference and push a number `2`. So that's how it works.

`[...x]` is the spread syntax, which returns items from `x,` then creates a new array from the items 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
